packages/ui/src/AccountSelector/index.tsx

```diff
@@ -0,0 +1,156 @@
+  gap: ${props => props.theme.spacing(2)};
+  padding: ${props => props.theme.spacing(2)} 0;
+`;
+
+const MAX_INDEX = 2 ** 32;
+
+interface ControlledProps {
+  /**
+   * The current account index.
+   *
+   * Leave this undefined if you want to use `<AccountSelector />` as an
+   * uncontrolled component.
+   */
+  value: number;
+  /**
+   * The account index the user is changing to.
+   *
+   * Leave this undefined if you want to use `<AccountSelector />` as an
+   * uncontrolled component.
+   */
+  onChange: (index: number) => void;
+}
+
+interface UncontrolledProps {
+  value?: undefined;
+  onChange?: undefined;
+}
+
+export type AccountSelectorProps = {
+  /**
+   * A possibly async function that, when given an address index (and a boolean
+   * indicating whether the address should be ephemeral), returns an `Address`.
+   *
+   * If left undefined, no address will be rendered.
+   */
+  getAddressByIndex?: (index: number, ephemeral: boolean) => Promise<Address> | Address;
+  /**
+   * An array of address indexes to switch between, if you want to limit the
+   * options. No need to sort them, as the component will do that for you.
+   */
+  filter?: number[];
+} & (ControlledProps | UncontrolledProps);
+
+/**
+ * Allows users to navigate between their accounts, either for viewing their
+ * addresses, or for toggling between accounts to control an account-based view.
+ *
+ * Can be used as either a controlled or uncontrolled component. To use as a
+ * controlled component, pass the `value` and `onChange` props. To use as an
+ * uncontrolled component (for when, e.g., you just want to enable a user to
+ * page through their account addresses, but doing so won't affect anything else
+ * on the page), leave those props undefined.
+ */
+export const AccountSelector = (props: AccountSelectorProps) => {
+  const {
+    index,
+    handleChange,
+    handleClickPrevious,
+    previousButtonEnabled,
+    handleClickNext,
+    nextButtonEnabled,
+    address,
+    ephemeral,
+    setEphemeral,
+    loading,
+  } = useAccountSelector(props);
+
+  return (
+    <Root>
+      <TextInput
+        type='number'
+        value={index.toString()}
+        min={0}
+        max={MAX_INDEX}
+        onChange={value => {
+          /**
+           * Don't allow manual account number entry when there's a
+           * filter.
+           *
+           * @todo: Change this to only call `handleChange()` when the
+           * user presses Enter? Then it could validate that the entered
+           * account index is in the filter.
+           */
+          if (props.filter) {
+            return;
+          }
+          const valueAsNumber = Number(value);
+          const valueLength = value.replace(/^0+/, '').length;
+
+          if (valueAsNumber > MAX_INDEX || valueLength > MAX_INDEX.toString().length) {
+            return;
+          }
+
+          handleChange(valueAsNumber);
+        }}
+        startAdornment={<StartAdornment>Account #</StartAdornment>}
```

Comment:
question: maybe it's worth making not `div` but `label` with `cursor: pointer`, so that users will have a larger hover area to select the input? So far it is still not obvious the AddressSelector is the input

Reply:
Great idea. I didn't go with your exact implementation, though: `AccountSelector` is a field that will often be nested inside `<FormField />`, which itself renders a `<label />`. Nesting `<label />`s is invalid HTML. So, to make this element usable inside or outside of a `<FormField />`, I've changed `<StartAdornment />` to be an (accessible) clickable div with a pointer cursor: https://github.com/penumbra-zone/web/pull/1682/commits/97f12855d5de48f324feeb87c5fb8c8439f56c0b